compiler/rustc_typeck/src/check/regionck.rs

```diff
@@ -771,7 +771,19 @@ impl<'a, 'tcx> RegionCtxt<'a, 'tcx> {
         debug!("link_upvar_region(borrorw_region={:?}, upvar_id={:?}", borrow_region, upvar_id);
         // A by-reference upvar can't be borrowed for longer than the
         // upvar is borrowed from the environment.
-        match self.typeck_results.borrow().upvar_capture(upvar_id) {
+        let closure_local_def_id = upvar_id.closure_expr_id;
+        let upvar_capture = self
+            .typeck_results
+            .borrow()
+            .closure_min_captures_flattened(closure_local_def_id.to_def_id())
+            .find(|captured_place| match captured_place.place.base {
+                PlaceBase::Upvar(id) => id == upvar_id,
```

Comment:
This feels a little incorrect because what this will do is stop at the first one.  What we want to do is create a new iterator using filter here instead of find.

And then iterate over this filtered iterator and do the match upvar_capture for each `captured_place` in that iterator

Reply:
Actually, on second thoughts you don't even need filter. 

`upvar_id.var_path.hir_id` gives u the root variable `var_hir_id`, and the closure def id to basically iterate over places captured starting at that root variable.

You can probably chain a bunch of things and do 

```rust
for captured_place in self.typeck_results.borrow().closure_min_captures(&closure_def_id).and_then(|min| min.get(&upvar_id.var_path.hir_id)).iter().flatten() {

}
```

```rust
self
.typeck_results
.borrow()
.closure_min_captures(&closure_def_id) // This returns Option<RootVariableCaptureList>
.and_then(|min| min.get(&upvar_id.var_path.hir_id)) // This returns Option<CaptureList>
.iter() // This returns Iterator<Item = CaptureList>
.flatten() // This returns Iterator<Item = CapturedPlace>
````